wlcreate/Start-Asian-Love
# Convert Wins and Wins ID Page to Sanity

## Description

Related Issue: #37 



We successfully rendered Wins Page and Wins ID page using `getStaticPaths` and `getStaticProps` to fetch to Sanity.  We also deleted the old Wins ID pages. 

Since there are many resources being displayed (with more to come) on the wins page, `getStaticProps` fetches the resources at build time. To make it easier to render and reuse the `WinsResources` component, after we receive all of the resources we organize them by the category they are (heritage, cultural, art, contributions, or representation). We are then passing them as props to Wins, destructuring props, and then passing again as props to the `WinsResources` component.

Similarly, our new `[slug].js` file/path is dynamically rendering all of the individual resources at build time. Therefore, instead of the route being `/wins/contributions/1`, it is now `/wins/[wins._id]`. As a result we no longer need to pass the category as props to `WinsResources`. 

Within `[slug].js` we use `getStaticPaths` to pre-render pages based on data and `getStaticProps` to fetch data at build time. Since in `WinsResources` we set `Link` to direct to `/wins/wins._id`, in `getStaticPaths` we are creating pages based on wins' and wins' `_id`, which are then returned and sent to `getStaticProps`. Then, based on the params and slug sent in `getStaticProps` we make a request to Sanity for the specific win that matches the path the user is trying to view. The win and preview of the win is then passed to Win as props and rendered on the page.

Note: We had some issue with the image rendering on the page; at first the image is null but once is fetched can be rendered. A quick fix we decided on was to check if the image is null/still fetching and changing what is displayed based on that.

Future Things to Think About:
- `isFallback`
- Loading component and/or what do we want to show when something is still being fetched (esp images)?
- Error 404 page
- Images are stretched out. Will need to edit the hot spots on Sanity studio and update images accordingly.
   - images for support are 1200px width and 900px height with a resolution of 300 (need to convert to webp)
   - images for wins are ???
- Wins ID URL is a long ID code. Might want to use something else for each individual win (ex: creating a "slug" attribute with "current" attached to it with another version of the resource's title)
- Using the `reusableButtons` component to render the categories on the page instead of hard-coding them

Resources:



- [Repo of Sanity demo by Sanity](https://github.com/sanity-io/next-recipe-app/blob/main/pages/recipes/%5Bslug%5D.js)
- [Repo of Sanity demo by Nextjs](https://github.com/vercel/next.js/blob/canary/examples/cms-sanity/pages/posts/%5Bslug%5D.js)
- [GROQ Cheat Sheet](https://www.sanity.io/docs/query-cheat-sheet)
- [What it means to have fallback true](https://nextjs.org/docs/basic-features/data-fetching#fallback-true)
- [Info of getStaticProps and getStaticPaths](https://nextjs.org/docs/basic-features/data-fetching)

## Acceptance Criteria



- [x]   Fetch and render wins resources database from Sanity.

## Self-check

Please check all that apply:




## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
| ✓  | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

## Updates

### Before

<img width="1764" alt="Screen Shot 2021-08-10 at 7 53 57 PM" src="https://user-images.githubusercontent.com/60532744/128949468-5c5a1ee7-5c44-479a-8588-9c82cd7050a4.png">

<img width="1371" alt="Screen Shot 2021-08-10 at 7 54 13 PM" src="https://user-images.githubusercontent.com/60532744/128949485-015fd4e4-5c74-406a-8f5f-e3df6b8e27ef.png">



### After

<img width="1797" alt="Screen Shot 2021-08-10 at 7 54 30 PM" src="https://user-images.githubusercontent.com/60532744/128949505-ecb55524-36e2-4820-bbac-7d9e9c1f1694.png">

<img width="1400" alt="Screen Shot 2021-08-10 at 7 54 45 PM" src="https://user-images.githubusercontent.com/60532744/128949517-e29cfd9b-be5d-4f59-9e87-a88338f25413.png">




## Testing Steps / QA Criteria

- `git pull`
- switch to `sanity-backend-wins` branch
- `npm run dev`
- Check to see if Wins page and Wins ID pages load
